### PR TITLE
Use metrics for preexecution tests

### PR DESCRIPTION
**Pull request by KTeofanidis**
Add new metric to count write requests that come from pre-execution. Use that as well as the preProcReqSentForFurtherProcessing metric in tests.

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -3352,6 +3352,7 @@ ReplicaImp::ReplicaImp(bool firstTime,
       metric_total_fastPath_{metrics_.RegisterCounter("totalFastPaths")},
       metric_total_slowPath_requests_{metrics_.RegisterCounter("totalSlowPathRequests")},
       metric_total_fastPath_requests_{metrics_.RegisterCounter("totalFastPathRequests")},
+      metric_total_preexec_requests_written_{metrics_.RegisterCounter("totalPreExecRequestsWritten")},
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:3355]: totalPreExecRequestsWritten-> totalPreExecRequestsExecuted
metric_total_preexec_requests_written_-> metric_total_preexec_requests_executed_

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 2]
```diff
@@ -271,6 +273,9 @@ void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequest
       preProcessorMetrics_.preProcReqIgnored.Get().Inc();
       return;
     }
+    // count requests that would have been ignored if they didn't come from non-primary replicas
+    if (clientId != senderId && myReplica_.isClientRequestInProcess(clientId, reqSeqNum))
+      preProcessorMetrics_.preProcReqForwardedNotIgnored.Get().Inc();
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:278]: preProcReqForwardedNotIgnored-> preProcReqForwardedByNonPrimaryNotIgnored

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 3]
```diff
@@ -271,6 +273,9 @@ void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequest
       preProcessorMetrics_.preProcReqIgnored.Get().Inc();
       return;
     }
+    // count requests that would have been ignored if they didn't come from non-primary replicas
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.cpp:276]: -> Count requests arrived through non-primary replicas and accepted in spite of the same request has been processing by the primary replica

**Review by yuliasherman — APPROVED** [review]
LGTM

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 4]
```diff
@@ -171,9 +171,11 @@ class PreProcessor {
     concordMetrics::CounterHandle preProcessRequestTimedout;
     concordMetrics::CounterHandle preProcReqSentForFurtherProcessing;
     concordMetrics::CounterHandle preProcPossiblePrimaryFaultDetected;
+    concordMetrics::CounterHandle preProcReqForwardedByNonPrimaryNotIgnored;
```
Comment (teoparvanov) [inline bftengine/src/preprocessor/PreProcessor.hpp:174]: Can we drop this metric, now that we have `metric_total_preexec_requests_executed_` ?

File: tests/apollo/util/bft.py [hunk 5]
```diff
@@ -1118,3 +1118,21 @@ async def do_key_exchange(self):
                 lastExecutedVal = await self.metrics.get(0, *lastExecutedKey)
             self.stop_all_replicas()
             return lastExecutedVal
+
+    async def assert_pre_proc_req_sent_for_further_processing(self, replica_id, num_requests):
```
Comment (teoparvanov) [inline tests/apollo/util/bft.py:1122]: `assert_successful_pre_executions_count()`

File: tests/apollo/util/bft.py [hunk 6]
```diff
@@ -1118,3 +1118,21 @@ async def do_key_exchange(self):
...
+        try:
+            pre_proc_req = 0
+            total_pre_exec_requests_executed = 0
+            with trio.fail_after(5):
+                while pre_proc_req < num_requests or \
+                        total_pre_exec_requests_executed != num_requests:
+                    key1 = ["preProcessor", "Counters", "preProcReqSentForFurtherProcessing"]
```
Comment (teoparvanov) [inline tests/apollo/util/bft.py:1129]: no need for this anymore, right?

**Review by teoparvanov — COMMENTED** [review]
Suggesting some simplifications...

File: tests/apollo/util/bft.py [hunk 7]
```diff
@@ -1118,3 +1118,21 @@ async def do_key_exchange(self):
...
+                    pre_proc_req = await self.metrics.get(replica_id, *key1)
+                    key2 = ["replica", "Counters", "totalPreExecRequestsExecuted"]
+                    total_pre_exec_requests_executed = await self.metrics.get(replica_id, *key2)
+                    await trio.sleep(1)
```
Comment (teoparvanov) [inline tests/apollo/util/bft.py:1133]: 0.1 ?

**Review by teoparvanov — APPROVED** [review]
LGTM

**Pull request merged**